File: leetcode/876middleoflinkedlist/another_answer.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        it_f, it_s = head, head
        while it_s and it_f and it_f.next:
            it_s = it_s.next
            it_f = it_f.next.next

        return it_s

File: leetcode/0141/another_answer.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (!head) return false;
        ListNode* slow = head, *fast = head->next;
        while (fast != NULL && fast->next != NULL && slow != fast) {
            fast = fast->next->next;
            slow = slow->next;
        }
        return fast == slow; 
    }
};
